jaxrs-api/src/main/java/jakarta/ws/rs/client/ClientBuilder.java

```diff
@@ -59,16 +61,31 @@ public static ClientBuilder newBuilder() {
         try {
             Object delegate = FactoryFinder.find(JAXRS_DEFAULT_CLIENT_BUILDER_PROPERTY, ClientBuilder.class);
             if (!(delegate instanceof ClientBuilder)) {
-                Class pClass = ClientBuilder.class;
+                Class<?> pClass = ClientBuilder.class;
                 String classnameAsResource = pClass.getName().replace('.', '/') + ".class";
-                ClassLoader loader = pClass.getClassLoader();
-                if (loader == null) {
-                    loader = ClassLoader.getSystemClassLoader();
+                final String errorMessage;
+                if (System.getSecurityManager() == null) {
+                    ClassLoader loader = pClass.getClassLoader();
+                    if (loader == null) {
+                        loader = ClassLoader.getSystemClassLoader();
+                    }
+                    URL targetTypeURL = loader.getResource(classnameAsResource);
+                    errorMessage = "ClassCastException: attempting to cast"
+                            + delegate.getClass().getClassLoader().getResource(classnameAsResource)
+                            + " to " + targetTypeURL;
+                } else {
+                    errorMessage = AccessController.doPrivileged((PrivilegedAction<String>) () -> {
+                        ClassLoader loader = pClass.getClassLoader();
```

Comment:
Maybe we should refactor into private method instead of duplicating code.

Reply:
Should be done if I've understood correctly :)